Refactor out shape test into InputMetadata::maybe_reduce

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119562
* __->__ #119559

I'm going to gut this function shortly, and having it all on
InputMetadata is convenient for this purpose.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Refactor out shape test into InputMetadata::maybe_expand

I'm going to gut this function shortly, and having it all on
InputMetadata is convenient for this purpose.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -850,14 +850,7 @@ void validate_outputs(
-        TORCH_CHECK(false, format_error(message.str()));
-      }
-    }
+    grad = metadata.maybe_expand(i, std::move(grad), format_error);
```
It should be maybe_reduce

albanD reviewed (commented):
The name is reversed but otherwise sounds fine.

ezyang committed:
Update on "Refactor out shape test into InputMetadata::maybe_expand"


I'm going to gut this function shortly, and having it all on
InputMetadata is convenient for this purpose.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
